Firestore/Source/Remote/FSTSerializerBeta.mm

```diff
@@ -572,31 +573,95 @@ - (FieldMask)decodedFieldMask:(GCFSDocumentMask *)fieldMask {
+  if (fieldTransform.transformation().type() == TransformOperation::Type::ServerTimestamp) {
+    proto.setToServerValue = GCFSDocumentTransform_FieldTransform_ServerValue_RequestTime;
+
+  } else if (fieldTransform.transformation().type() == TransformOperation::Type::ArrayUnion) {
+    auto array_transform = static_cast<const ArrayTransform &>(fieldTransform.transformation());
+    proto.appendMissingElements = [self encodedArrayTransformElements:array_transform.elements()];
+
+  } else if (fieldTransform.transformation().type() == TransformOperation::Type::ArrayRemove) {
+    auto array_transform = static_cast<const ArrayTransform &>(fieldTransform.transformation());
+    proto.removeAllFromArray_p = [self encodedArrayTransformElements:array_transform.elements()];
+
+  } else {
+    FSTFail(@"Unknown transform: %d type", fieldTransform.transformation().type());
+  }
+  return proto;
+}
+
+- (GCFSArrayValue *)encodedArrayTransformElements:(const std::vector<FSTFieldValue *> &)elements {
+  GCFSArrayValue *proto = [GCFSArrayValue message];
+  NSMutableArray<GCFSValue *> *protoContents = [proto valuesArray];
+
+  for (FSTFieldValue *element : elements) {
+    GCFSValue *converted = [self encodedFieldValue:element];
+    [protoContents addObject:converted];
+  }
+  return proto;
+}
+
 - (std::vector<FieldTransform>)decodedFieldTransforms:
     (NSArray<GCFSDocumentTransform_FieldTransform *> *)protos {
   std::vector<FieldTransform> fieldTransforms;
   fieldTransforms.reserve(protos.count);
+
   for (GCFSDocumentTransform_FieldTransform *proto in protos) {
-    FSTAssert(
-        proto.setToServerValue == GCFSDocumentTransform_FieldTransform_ServerValue_RequestTime,
-        @"Unknown transform setToServerValue: %d", proto.setToServerValue);
-    fieldTransforms.emplace_back(
-        FieldPath::FromServerFormat(util::MakeStringView(proto.fieldPath)),
-        absl::make_unique<ServerTimestampTransform>(ServerTimestampTransform::Get()));
+    switch (proto.transformTypeOneOfCase) {
+      case GCFSDocumentTransform_FieldTransform_TransformType_OneOfCase_SetToServerValue: {
+        FSTAssert(
+            proto.setToServerValue == GCFSDocumentTransform_FieldTransform_ServerValue_RequestTime,
+            @"Unknown transform setToServerValue: %d", proto.setToServerValue);
+        fieldTransforms.emplace_back(
+            FieldPath::FromServerFormat(util::MakeStringView(proto.fieldPath)),
+            absl::make_unique<ServerTimestampTransform>(ServerTimestampTransform::Get()));
+        break;
+      }
+
+      case GCFSDocumentTransform_FieldTransform_TransformType_OneOfCase_AppendMissingElements: {
+        std::vector<FSTFieldValue *> elements =
+            [self decodedArrayTransformElements:proto.appendMissingElements];
+        fieldTransforms.emplace_back(
+            FieldPath::FromServerFormat(util::MakeStringView(proto.fieldPath)),
+            absl::make_unique<ArrayTransform>(TransformOperation::Type::ArrayUnion, elements));
```

Comment:
Okay, this is almost done however I just noticed that the ArrayTransform constructor is busted:

  * It's also passing `const foo` when it should either be `const foo&` or just `foo`
  * It's not moving into `elements_` which means we're suffering an extra copy
  * `const Type&` is unnecessary for trivial, small types. Just `Type`

Taken together, please make it:

```c++
class ArrayTransform : public TransformOperation {
 public:
  ArrayTransform(Type type, std::vector<FSTFieldValue*> elements)
      : type_(type), elements_(std::move(elements)) {
  }
```

Reply:
Good call, done.

Fun fact, I was curious what would happen if I tried to take elements as a const reference and still use std::move() on it, i.e.:
```
  ArrayTransform(Type type, const std::vector<FSTFieldValue*> &elements)
      : type_(type), elements_(std::move(elements)) {
  }
```

I expected a compiler error because it seems unsafe to use `std::move()` on a const object but it compiles fine and seems to somehow end up just invoking the `std::vector<>` copy constructor and both my `std::move()` calls (here and in the caller) are apparently elided (`std::vector<>`'s move constructor is not called at all)...  So I guess `std::move()` is more of a compiler hint than an actual deterministic thing... wacky.